addons/hr_timesheet/models/project.py

```diff
@@ -44,14 +46,30 @@ def _check_allow_timesheet(self):
 
     @api.depends('timesheet_ids')
     def _compute_total_timesheet_time(self):
+        timesheets_read_group = self.env['account.analytic.line'].read_group(
+            [('project_id', 'in', self.ids)],
+            ['project_id', 'unit_amount', 'product_uom_id'],
+            ['project_id', 'product_uom_id'],
+            lazy=False)
+        timesheet_time_dict = defaultdict(list)
+        uom_ids = set(self.timesheet_encode_uom_id.ids)
+        for result in timesheets_read_group:
+            uom_id = result['product_uom_id'] and result['product_uom_id'][0]
+            if uom_id:
+                uom_ids.add(uom_id)
+            timesheet_time_dict[result['project_id'][0]].append((uom_id, result['unit_amount']))
+        uoms_dict = {uom.id: uom for uom in self.env['uom.uom'].search([('id', 'in', list(uom_ids))])}
         for project in self:
-            total_time = 0.0
-            for timesheet in project.timesheet_ids:
-                # Timesheets may be stored in a different unit of measure, so first
-                # we convert all of them to the reference unit
-                total_time += timesheet.unit_amount * timesheet.product_uom_id.factor_inv
+            project_uom = uoms_dict[project.timesheet_encode_uom_id.id]
+            # Timesheets may be stored in a different unit of measure, so first
+            # we convert all of them to the reference unit
+            # if the timesheet has no product_uom_id then we take the one of the project
+            total_time = sum([
+                unit_amount * (uoms_dict[product_uom_id] if product_uom_id else project_uom).factor_inv
+                for product_uom_id, unit_amount in timesheet_time_dict[project.id]
+            ], 0.0)
```

Comment:
We could use the *generator expression* rather than the list comprehension to not store the entire list in memory since we just need the sum of the list values. What do you think ? 🙂
With the *generator expression* would have this:
```suggestion
            total_time = sum((
                unit_amount * (uoms_dict[product_uom_id] if product_uom_id else project_uom).factor_inv
                for product_uom_id, unit_amount in timesheet_time_dict[project.id]
            ), 0.0)
```

Reply:
I like the idea 🙂